pyclam/criterion.py

```diff
@@ -119,6 +141,90 @@ def __call__(self, root: Cluster) -> Set[Cluster]:
         return {cluster for cluster in root.manifold.layers[-1].clusters}
 
 
+class Layer(SelectionCriterion):
+    def __init__(self, depth: int):
+        if depth <= 0:
+            raise ValueError(f'expected a positive depth. got: {depth}')
+        self.depth = depth
+
+    def __call__(self, root: Cluster):
+        manifold: Manifold = root.manifold
+        if manifold.depth <= self.depth:
+            return [cluster for cluster in manifold.layers[-1].clusters]
+        else:
+            return [cluster for cluster in manifold.layers[self.depth].clusters]
+
+
+class SelectionClauses(SelectionCriterion):
+    def __init__(self, clauses: Iterable[Clause]):
+        if any((not isinstance(c, Clause) for c in clauses)):
+            raise TypeError(f'selection clauses mut all be of type Clause.')
+        self.clauses: List[Clause] = list(clauses)
+
+    def __call__(self, root: Cluster) -> Set[Cluster]:
+        logging.info(f'selecting clusters for optimal graph...')
+        manifold = root.manifold
+        clusters: Set[Cluster] = {cluster for cluster in manifold.layers[1].clusters}
+        selected: Set[Cluster] = set()
+
+        while clusters:
+            new_clusters: Set[Cluster] = set()
+            for cluster in clusters:
+                if cluster.children:
+                    for clause in self.clauses:
+                        if clause(cluster):
+                            selected.add(cluster)
+                            break
+                    else:
+                        new_clusters.update(cluster.children)
+                else:
+                    selected.add(cluster)
+            clusters = new_clusters
+
+        logging.info(f'selected {len(selected)} clusters for optimal graph')
+        return selected
+
+
+class LinearRegressionConstants(SelectionCriterion):
+    def __init__(self, constants: Union[np.array, List[float]]):
+        constants = np.asarray(constants, dtype=float)
+        if constants.shape != (6,):
+            raise ValueError(f'expected a vector of 6 elements. Got {constants.shape} instead.')
```

Comment:
I think 6 here should be a constant

Reply:
we have been unable to come up with more features to use for the meta-ml models. If we ever do, we will change 6 as needed